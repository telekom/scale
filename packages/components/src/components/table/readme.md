# scale-table



<!-- Auto Generated Below -->


## Usage

### Table

### Default

<scale-table>
  <table>
    <thead>
    <tr>
      <th>Title</th>
      <th>Stats</th>
    </tr>
    </thead>
    <tbody>
    <tr>
      <td>Jane</td>
      <td>9.356</td>
    </tr>
    <tr>
      <td>Jack</td>
      <td>3.356</td>
    </tr>
    <tr>
      <td>John</td>
      <td>3.356</td>
    </tr>
    </tbody>
    <tfoot>
    <tr>
      <td>Total</td>
      <td>100.245,10</td>
    </tr>
    </tfoot>
  </table>
</scale-table>

### Small size

<scale-table size="small">
  <table>
    <thead>
    <tr>
      <th>Title</th>
      <th>Stats</th>
    </tr>
    </thead>
    <tbody>
    <tr>
      <td>Jane</td>
      <td>9.356</td>
    </tr>
    <tr>
      <td>Jack</td>
      <td>3.356</td>
    </tr>
    <tr>
      <td>John</td>
      <td>3.356</td>
    </tr>
    </tbody>
    <tfoot>
    <tr>
      <td>Total</td>
      <td>100.245,10</td>
    </tr>
    </tfoot>
  </table>
</scale-table>

### Full sorting example

<scale-table show-sort="true">
  <table id="sortable-table">
    <div style="padding: 20px 8px 0px 8px;">
      <scale-text size="h5">Table title</scale-text>
    </div>
    <thead>
    <tr>
      <!--  Pay attention to the usage of `th` `aria-sort` attribute-->
      <!--  which supports all the native values: `descending`, `ascending`, `none`, `other`.-->
      <th aria-sort="ascending" onclick="sortTable(this)">Title</th>
      <!--  When you want to disable sorting for a particular column,-->
      <!--  add the `aria-disabled` attribute to the `th`.-->
      <th aria-disabled="true" onclick="sortTable(this)">Tags</th>
      <th onclick="sortTable(this)">Stats</th>
      <th onclick="sortTable(this)" style="text-align: right;">Time</th>
      <th onclick="sortTable(this)" style="text-align: right;">
        Euros
      </th>
    </tr>
    </thead>
    <tbody>
    <tr>
      <td>Jane</td>
      <td>
        <scale-tag size="small" style="margin-right: 8px"
        >Other</scale-tag
        >
        <scale-tag size="small" style="margin-right: 8px"
        >N/A</scale-tag
        >
        <scale-tag size="small">Demo</scale-tag>
      </td>
      <td>
        <div style="display: inline-flex; align-items: center;">
          <span style="margin-right: 8px">9.356</span>
          <scale-progress-bar
                  size="info"
                  style=" max-width: 120px;"
                  stroke-width="6"
                  percentage="90"
          />
        </div>
      </td>
      <td style="text-align: right; font-weight: 100;">00:00:20</td>
      <td style="text-align: right; font-weight: 100;">100.245,10</td>
    </tr>
    <tr>
      <td>Jack</td>
      <td>
        <scale-tag size="small" style="margin-right: 8px"
        >Other</scale-tag
        >
        <scale-tag size="small" style="margin-right: 8px"
        >N/A</scale-tag
        >
        <scale-tag size="small">Demo</scale-tag>
      </td>
      <td>
        <div style="display: inline-flex; align-items: center;">
          <span style="margin-right: 8px">3.356</span>
          <scale-progress-bar
                  size="info"
                  style=" max-width: 120px;"
                  stroke-width="6"
                  percentage="90"
          />
        </div>
      </td>
      <td style="text-align: right; font-weight: 100;">00:00:30</td>
      <td style="text-align: right; font-weight: 100;">100.345,10</td>
    </tr>
    <tr>
      <td>John</td>
      <td>
        <scale-tag size="small" style="margin-right: 8px"
        >Other</scale-tag
        >
        <scale-tag size="small" style="margin-right: 8px"
        >N/A</scale-tag
        >
        <scale-tag size="small">Demo</scale-tag>
      </td>
      <td>
        <div style="display: inline-flex; align-items: center;">
          <span style="margin-right: 8px">6.356</span>
          <scale-progress-bar
                  size="info"
                  style=" max-width: 120px;"
                  stroke-width="6"
                  percentage="90"
          />
        </div>
      </td>
      <td style="text-align: right; font-weight: 100;">00:00:40</td>
      <td style="text-align: right; font-weight: 100;">100.445,10</td>
    </tr>
    </tbody>
    <tfoot>
    <tr>
      <td>Total</td>
      <td />
      <td />
      <td style="text-align: right;">00:00:20</td>
      <td style="text-align: right;">100.245,10</td>
    </tr>
    </tfoot>
  </table>
  <script>
    function getNextSort(sort) {
      if (!sort || ["none", "other"].includes(sort)) {
        return "ascending";
      }
      if (sort === "ascending") {
        return "descending";
      }
      return "none";
    };
    
    function sortTable(th) {
      const currentSort = th.getAttribute("aria-sort");
      const nextSort = getNextSort(currentSort);
      const tableHeaders = Array.from(document.getElementsByTagName("TH"));
      const columnIndex = tableHeaders.findIndex((x) => x === th);
    
      // clean up previous aria-sort value
      tableHeaders.forEach((tableHeader) => {
        tableHeader.setAttribute("aria-sort", "none");
      });
    
      // set actual sort
      th.setAttribute("aria-sort", nextSort);
    
      // Taken from:
      // https://www.w3schools.com/howto/tryit.asp?filename=tryhow_js_sort_table
      var table, rows, switching, i, x, y, shouldSwitch;
      table = document.getElementById("sortable-table");
      switching = true;
      /*Make a loop that will continue until
        no switching has been done:*/
      while (switching) {
        //start by saying: no switching is done:
        switching = false;
        rows = table.rows;
        /*Loop through all table rows (except the
            first, which contains table headers):*/
        for (i = 1; i < rows.length - 1; i++) {
          //start by saying there should be no switching:
          shouldSwitch = false;
    
          /*Get the two elements you want to compare,
                one from current row and one from the next:*/
          x = rows[i].getElementsByTagName("TD")[columnIndex];
          y = rows[i + 1].getElementsByTagName("TD")[columnIndex];
    
          if (
            !["descending", "ascending"].includes(nextSort) ||
            x.parentElement.parentElement.tagName === "TFOOT" ||
            y.parentElement.parentElement.tagName === "TFOOT"
          ) {
            break;
          }
    
          //check if the two rows should switch place:
          if (
            (nextSort === "ascending" ? y : x).innerHTML.toLowerCase() >
            (nextSort === "ascending" ? x : y).innerHTML.toLowerCase()
          ) {
            //if so, mark as a switch and break the loop:
            shouldSwitch = true;
            break;
          }
        }
        if (shouldSwitch) {
          /*If a switch has been marked, make the switch
                and mark that a switch has been done:*/
          rows[i].parentNode.insertBefore(rows[i + 1], rows[i]);
          switching = true;
        }
      }
    }
  </script>
</scale-table>



## Properties

| Property   | Attribute   | Description                           | Type      | Default     |
| ---------- | ----------- | ------------------------------------- | --------- | ----------- |
| `showSort` | `show-sort` | (optional) Display sort arrows on/off | `boolean` | `false`     |
| `size`     | `size`      | (optional) Visual size                | `string`  | `'default'` |
| `styles`   | `styles`    | (optional) Injected CSS styles        | `string`  | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
