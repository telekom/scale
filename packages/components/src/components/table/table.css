/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

scale-table {
  --radius: var(--telekom-radius-standard) var(--telekom-radius-standard) 0 0;
  /* FIXME:TOKEN previously --scl-color-background-light */
  --background: var(--telekom-color-ui-subtle);
  --color: var(--telekom-color-text-and-icon-standard);
  --color-hover: var(--telekom-color-text-and-icon-primary-hovered);
  --color-active: var(--telekom-color-text-and-icon-primary-pressed);
  --font-size: var(--telekom-typography-font-size-small);
  --font-weight: var(--telekom-typography-font-weight-bold);

  --spacing-tbody-td: var(--telekom-spacing-unit-x4)
    var(--telekom-spacing-unit-x2);
  /*FIXME:TOKEN border color should be different from background color running out of choice? */
  /* FIXME:TOKEN previously --scl-color-grey-20 */
  --border-bottom-tbody-td: var(--telekom-spacing-unit-x025) solid
    var(--telekom-color-ui-subtle);
  --background-tbody: var(--telekom-color-ui-base);

  /* FIXME:TOKEN hardcoded values #ededed */
  --background-tbody-tr-hover: var(--telekom-color-ui-subtle);
  --background-tfoot: var(--telekom-color-ui-base);
  --border-bottom-tfoot-td: var(--telekom-spacing-unit-x025) solid
    var(--telekom-color-ui-extra-strong);

  --spacing-th-sortable: 0 var(--telekom-spacing-unit-x2) 0 0;
  /* FIXME:TOKEN hardcoded values #ededed */
  --background-th-sortable-hover: var(--telekom-color-ui-subtle);
  --background-th-sortable-active: var(--telekom-color-ui-light);
  --box-shadow-th-sortable-focus: inset 0 0 0 var(--telekom-spacing-unit-x05)
    var(--telekom-color-functional-focus);
  /*FIXME:TOKEN --scl-color-grey-0? */
  --background-tr-striped: var(--scl-color-grey-0);
  --padding: var(--telekom-spacing-unit-x0) var(--telekom-spacing-unit-x2);
}
.table {
  display: block;
  overflow: auto;
  border-radius: var(--radius);
  background: var(--background);
  color: var(--color);
}
.table table {
  width: 100%;
  white-space: nowrap;
  border-spacing: 0;
  border-collapse: collapse;
}
.table th {
  padding: var(--padding);
  text-align: left;
  line-height: 32px; /*  TODO: revise, this is weird */
  color: var(--color);
  font-size: var(--font-size);
}
.table td {
  padding: var(--padding);
}
.table tbody tr td {
  padding: 8px;
}
.table tbody td {
  border-bottom: var(--border-bottom-tbody-td);
}
.table thead tr th:last-of-type,
.table tbody tr td:last-of-type,
.table tfoot tr td:last-of-type {
  padding-right: 12px;
}

.table tfoot {
  background: var(--background-tfoot);
}
.table tbody {
  background: var(--background-tbody);
}
.table tfoot tr td {
  padding: 8px;
  font-weight: var(--font-weight);
  border-bottom: var(--border-bottom-tfoot-td);
}
.table .scale-sort-indicator svg {
  display: none;
}
.table tbody tr:hover {
  background: var(--background-tbody-tr-hover);
}
.table th:focus {
  outline: none;
}
.table--size-default {
}
.table--size-default tbody tr td {
  padding: var(--spacing-tbody-td);
}
.table--sortable {
}
.table--sortable th {
  cursor: pointer;
  padding: var(--spacing-th-sortable);
  border-radius: var(--radius);
}
.table--sortable th:hover {
  padding: var(--spacing-th-sortable);
  background: var(--background-th-sortable-hover);
}
.table--sortable th:active {
  background: var(--background-th-sortable-active);
}
.table--sortable th:focus {
  box-shadow: var(--box-shadow-th-sortable-focus);
}
.table--sortable th .scale-sort-indicator svg {
  display: block;
}
.table--sortable th[aria-disabled] {
  padding: var(--padding);
  pointer-events: none;
}
.table--sortable th:not([aria-sort]) {
  padding: var(--padding);
}
.table--sortable th[aria-sort='none'] {
  padding: var(--padding);
}
.table--sortable
  th[aria-sort='ascending']
  .scale-sort-indicator
  polygon:first-of-type {
  fill: var(--color);
}
.table--sortable th:hover .scale-sort-indicator polygon:first-of-type {
  fill: var(--color-hover);
}
.table--sortable
  th:hover[aria-sort='ascending']
  .scale-sort-indicator
  polygon:first-of-type {
  fill: var(--color-hover);
}
.table--sortable
  th:active[aria-sort='ascending']
  .scale-sort-indicator
  polygon:first-of-type {
  fill: var(--color-active);
}
.table--sortable th[aria-sort='ascending'] .scale-sort-indicator polygon {
  fill: transparent;
}
.table--sortable th:hover[aria-sort='ascending'] .scale-sort-indicator polygon {
  fill: transparent;
}
.table--sortable
  th:active[aria-sort='ascending']
  .scale-sort-indicator
  polygon {
  fill: transparent;
}
.table--sortable
  th[aria-sort='descending']
  .scale-sort-indicator
  polygon:first-of-type {
  fill: transparent;
}
.table--sortable
  th:hover[aria-sort='descending']
  .scale-sort-indicator
  polygon:first-of-type {
  fill: transparent;
}
.table--sortable
  th:active[aria-sort='descending']
  .scale-sort-indicator
  polygon:first-of-type {
  fill: transparent;
}
.table--sortable th[aria-sort='descending'] .scale-sort-indicator polygon {
  fill: var(--color);
}
.table--sortable
  th:hover[aria-sort='descending']
  .scale-sort-indicator
  polygon {
  fill: var(--color-hover);
}
.table--sortable
  th:active[aria-sort='descending']
  .scale-sort-indicator
  polygon {
  fill: var(--color-active);
}
.table--sortable th .scale-sort-indicator {
  width: 16px;
  height: 16px;
  margin: 0 0 0 4px;
  display: inline-block;
  position: relative;
}
.table--sortable th .scale-sort-indicator > svg {
  top: 4px;
  left: 0;
  width: 16px;
  height: 16px;
  position: absolute;
}
.table--sortable th .scale-sort-indicator polygon {
  fill: transparent;
}
.table--sortable th[aria-sort='none']:hover {
  padding: 0 8px 0 0;
}
.table--sortable th[aria-sort='none'] .scale-sort-indicator {
  display: none;
}
.table--sortable th:not([aria-sort]):hover {
  padding: 0 8px 0 0;
}
.table--sortable th:not([aria-sort]) .scale-sort-indicator {
  display: none;
}
.table--sortable th:hover .scale-sort-indicator {
  display: inline-block;
}
.table--striped table tr:nth-child(even) {
  background: var(--background-tr-striped);
}
