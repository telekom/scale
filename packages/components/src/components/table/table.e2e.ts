import { newE2EPage } from '@stencil/core/testing';

describe('scale-table', () => {
  it('should match snapshot', async () => {
    const page = await newE2EPage();
    await page.setContent(`
      <scale-table>
        <div>Table header</div>
        <table>
          <thead>
            <tr>
              <th>Title</th>
              <th>Time</th>
              <th>Euros</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
          </tbody>
          <tfoot>
            <tr>
              <td>Total</td>
              <td />
              <td />
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
          </tfoot>
        </table>
      </scale-table>
    `);
    const element = await page.find('scale-table');
    expect(element).toHaveClass('hydrated');
  });
});
