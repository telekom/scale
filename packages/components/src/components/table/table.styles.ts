import { JssStyle } from 'jss';

export const styles: JssStyle = {
  table: {
    display: 'block',
    overflow: 'auto',
    borderRadius: '4px 4px 0 0',
    background: ({ colors }) => colors.grey[100],

    '& table': {
      fontFamily: ({ typography }) => typography.fontFamily,
      borderSpacing: 0,
      borderCollapse: 'collapse',
      width: '100%',
      whiteSpace: 'nowrap',
    },

    '& th': {
      color: ({ colors }) => colors.grey[600],
      lineHeight: '32px',
      fontSize: 12,
      textAlign: 'left',
      userSelect: 'none',
    },
    '& th, td': {
      padding: '0 8px',
    },

    '& tbody tr td': {
      padding: '8px',
    },

    '& tfoot, tbody': {
      background: '#FFFFFF',
    },

    '& tfoot tr td': {
      fontWeight: 600,
      padding: '8px',
      borderBottom: '1px solid #000000',
    },

    '& .scale-sort-indicator svg': {
      display: 'none',
    },
  },

  'table--size-default': {
    '& tbody tr td': {
      padding: '16px 8px',
    },
  },

  'table--sortable': {
    '& th': {
      cursor: 'pointer',

      '& .scale-sort-indicator svg': {
        display: 'block',
      },
      '&[aria-disabled="true"]': {
        pointerEvents: 'none',
      },
      '&[aria-sort="descending"] .scale-sort-indicator polygon:first-of-type': {
        fill: '#000000',
      },
      '&[aria-sort="descending"] .scale-sort-indicator polygon': {
        fill: '#DCDCDC',
      },
      '&[aria-sort="ascending"] .scale-sort-indicator polygon:first-of-type': {
        fill: '#DCDCDC',
      },
      '&[aria-sort="ascending"] .scale-sort-indicator polygon': {
        fill: '#000000',
      },
      '& .scale-sort-indicator': {
        left: '-8px',
        position: 'relative',
        display: 'inline-block',
        width: '12px',
        height: '20px',
        '& > svg': {
          position: 'absolute',
          left: 0,
          top: '4px',
          height: '24px',
          width: '24px',
        },
        '& polygon': {
          fill: '#DCDCDC',
        },
      },
    },
  },
};
