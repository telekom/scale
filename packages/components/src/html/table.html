<!DOCTYPE html>
<html dir="ltr" lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0"
    />
    <title>Stencil Component Starter</title>

    <script type="module" src="/build/scale-components.esm.js"></script>
    <link rel="stylesheet" href="/build/scale-components.css" />
  </head>

  <body>
    <h3>Table</h3>

    <script>
      const getNextSort = (sort) => {
        if (!sort || ['none', 'other'].includes(sort)) {
          return 'descending';
        }
        if (sort === 'descending') {
          return 'ascending';
        }
        return 'none';
      };

      function sortTable(th) {
        const currentSort = th.getAttribute('aria-sort');
        const nextSort = getNextSort(currentSort);
        const tableHeaders = Array.from(document.getElementsByTagName('TH'));
        const columnIndex = tableHeaders.findIndex((x) => x === th);

        // clean up previous aria-sort value
        tableHeaders.forEach((tableHeader) => {
          tableHeader.setAttribute('aria-sort', 'none');
        });

        // set actual sort
        th.setAttribute('aria-sort', nextSort);

        // Taken from:
        // https://www.w3schools.com/howto/tryit.asp?filename=tryhow_js_sort_table
        var table, rows, switching, i, x, y, shouldSwitch;
        table = document.getElementById('sortable-table');
        switching = true;
        /*Make a loop that will continue until
  no switching has been done:*/
        while (switching) {
          //start by saying: no switching is done:
          switching = false;
          rows = table.rows;
          /*Loop through all table rows (except the
        first, which contains table headers):*/
          for (i = 1; i < rows.length - 1; i++) {
            //start by saying there should be no switching:
            shouldSwitch = false;

            /*Get the two elements you want to compare,
              one from current row and one from the next:*/
            x = rows[i].getElementsByTagName('TD')[columnIndex];
            y = rows[i + 1].getElementsByTagName('TD')[columnIndex];

            if (
              !['ascending', 'descending'].includes(nextSort) ||
              x.parentElement.parentElement.tagName === 'TFOOT' ||
              y.parentElement.parentElement.tagName === 'TFOOT'
            ) {
              break;
            }

            //check if the two rows should switch place:
            if (
              (nextSort === 'descending' ? y : x).innerHTML.toLowerCase() >
              (nextSort === 'descending' ? x : y).innerHTML.toLowerCase()
            ) {
              //if so, mark as a switch and break the loop:
              shouldSwitch = true;
              break;
            }
          }
          if (shouldSwitch) {
            /*If a switch has been marked, make the switch
              and mark that a switch has been done:*/
            rows[i].parentNode.insertBefore(rows[i + 1], rows[i]);
            switching = true;
          }
        }
      }
      function handleKeyDown(e, th) {
        if (['Space', 'Enter'].includes(e.code)) {
          sortTable(th);
        }
      }
    </script>

    <scale-table size="default" show-sort="true">
      <table id="sortable-table">
        <thead>
          <tr>
            <th
              aria-sort="descending"
              onclick="sortTable(this)"
              onkeydown="handleKeyDown(event, this)"
              tabindex="0"
            >
              Title
            </th>
            <th aria-disabled="true">Tags</th>
            <th
              onclick="sortTable(this)"
              onkeydown="handleKeyDown(event, this)"
              tabindex="0"
            >
              Stats
            </th>
            <th
              onclick="sortTable(this)"
              onkeydown="handleKeyDown(event, this)"
              style="text-align: right"
              tabindex="0"
            >
              Time
            </th>
            <th
              onclick="sortTable(this)"
              onkeydown="handleKeyDown(event, this)"
              style="text-align: right"
              tabindex="0"
            >
              Euros
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Jane</td>
            <td>
              <scale-tag size="small" style="margin-right: 8px"
                >Other</scale-tag
              >
              <scale-tag size="small" style="margin-right: 8px">N/A</scale-tag>
              <scale-tag size="small">Demo</scale-tag>
            </td>
            <td>
              <div style="display: inline-flex; align-items: center">
                <span style="margin-right: 8px">9.356</span>
                <scale-progress-bar
                  variant="info"
                  style="max-width: 120px"
                  stroke-width="6"
                  percentage="90"
                />
              </div>
            </td>
            <td style="text-align: right; font-weight: 100">00:00:20</td>
            <td style="text-align: right; font-weight: 100">100.245,10</td>
          </tr>
          <tr>
            <td>Jack</td>
            <td>
              <scale-tag size="small" style="margin-right: 8px"
                >Other</scale-tag
              >
              <scale-tag size="small" style="margin-right: 8px">N/A</scale-tag>
              <scale-tag size="small">Demo</scale-tag>
            </td>
            <td>
              <div style="display: inline-flex; align-items: center">
                <span style="margin-right: 8px">3.356</span>
                <scale-progress-bar
                  variant="info"
                  style="max-width: 120px"
                  stroke-width="6"
                  percentage="90"
                />
              </div>
            </td>
            <td style="text-align: right; font-weight: 100">00:00:30</td>
            <td style="text-align: right; font-weight: 100">100.345,10</td>
          </tr>
          <tr>
            <td>John</td>
            <td>
              <scale-tag size="small" style="margin-right: 8px"
                >Other</scale-tag
              >
              <scale-tag size="small" style="margin-right: 8px">N/A</scale-tag>
              <scale-tag size="small">Demo</scale-tag>
            </td>
            <td>
              <div style="display: inline-flex; align-items: center">
                <span style="margin-right: 8px">6.356</span>
                <scale-progress-bar
                  variant="info"
                  style="max-width: 120px"
                  stroke-width="6"
                  percentage="90"
                />
              </div>
            </td>
            <td style="text-align: right; font-weight: 100">00:00:40</td>
            <td style="text-align: right; font-weight: 100">100.445,10</td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td>Total</td>
            <td />
            <td />
            <td style="text-align: right">00:00:20</td>
            <td style="text-align: right">100.245,10</td>
          </tr>
        </tfoot>
      </table>
    </scale-table>
  </body>
</html>
