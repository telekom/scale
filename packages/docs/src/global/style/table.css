
table {
  width: 100%;
  overflow: auto;
  margin-top: 28px;
  margin-bottom: 18px;
  border-spacing: 0;
  border-collapse: collapse;
  overflow-x: auto;
}

tr {
  background-color: #fff;
  border-top: 1px solid #c6cbd1;
}

th {
  font-size: 13px;
  font-weight: bold;
}

th,
td {
  padding: 10px 10px;
  border: 1px solid #dfe2e5;
}

td {
  border: 1px solid #dfe2e5;
  vertical-align: top;
  font-size: 14px;
}

td {
  line-height: 1.5;
}

table code {
  font-size: 13px;
}

tr:nth-child(2n) {
  background-color: #f6f8fa;
}


@media screen and (max-width: 590px) {
  table {
    display: block;
  }
}
