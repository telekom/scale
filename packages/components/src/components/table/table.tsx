import { Component, Prop, h, Host, Element } from '@stencil/core';
import { CssClassMap } from '../../utils/utils';
import classNames from 'classnames';
import { styles } from './table.styles';
import { CssInJs } from '../../utils/css-in-js';
import { StyleSheet } from 'jss';
import Base from '../../utils/base-interface';

@Component({
  tag: 'scale-table',
  shadow: false,
})
export class Table implements Base {
  @Element() hostElement: HTMLElement;
  /** (optional) Tag class */
  @Prop() customClass?: string = '';
  /** (optional) Display sort arrows on/off */
  @Prop() showSort?: boolean = false;
  /** (optional) Visual variant */
  @Prop() variant?: 'regular' | 'compressed' = 'regular';
  /** (optional) Injected jss styles */
  @Prop() styles?: StyleSheet;
  /** decorator Jss stylesheet */
  @CssInJs('Table', styles) stylesheet: StyleSheet;
  /** object of the slots in use */
  slots: { header?: Element; table?: Element } = {};

  componentWillUpdate() {}
  componentDidUnload() {}

  render() {
    // on initial render
    if (!this.slots.table) {
      // build object of slots
      // @ts-ignore - fromEntries should be fine here
      this.slots = Object.fromEntries(
        Array.from(this.hostElement.children).map(child => [child.slot, child])
      );

      // when showSort is enabled insert indicator arrows for each th of the found table
      if (this.showSort && this.slots.table) {
        this.slots.table.querySelectorAll('th').forEach(th => {
          th.insertAdjacentHTML(
            'afterbegin',
            `
            <span class="scale-sort-indicator">
              <svg width="24px" height="24px" viewBox="0 0 24 24">
                <polygon points="11.8284271 16.6568542 14.6568542 13.8284271 9 13.8284271" />
                <polygon points="11.8284271 8 14.6568542 10.8284271 9 10.8284271" />
              </svg>
            </span>`
          );
        });
      }

      // append header slot to thead
      if (this.slots.header) {
        this.slots.table
          .querySelector('thead')
          .insertAdjacentElement('afterbegin', this.slots.header);
      }
    }

    return (
      <Host class={this.getCssClassMap()}>
        <style>{this.stylesheet.toString()}</style>
        <slot name="table" />
      </Host>
    );
  }

  getCssClassMap(): CssClassMap {
    const { classes } = this.stylesheet;
    return classNames(
      classes.table,
      this.customClass && this.customClass,
      this.variant && classes[`table--variant-${this.variant}`],
      this.showSort && classes[`table--sortable`]
    );
  }
}
