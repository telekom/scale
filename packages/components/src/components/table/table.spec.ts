import { newSpecPage } from '@stencil/core/testing';
import { Table } from './table';
import { styles } from './table.styles';
import jss from 'jss';

describe('Table', () => {
  let element;
  let stylesheet;

  beforeEach(async () => {
    element = new Table();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
  });

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Table],
      html: `
      <scale-table>
        <div>Table header</div>
        <table>
          <thead>
            <tr>
              <th>Title</th>
              <th>Time</th>
              <th>Euros</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>University of Plymouth</td>
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
          </tbody>
          <tfoot>
            <tr>
              <td>Total</td>
              <td />
              <td />
              <td>00:00:20</td>
              <td>100.245,10</td>
            </tr>
          </tfoot>
        </table>
      </scale-table>
      `,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.size = 'big';
    stylesheet.addRule('table--size-big', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['table--size-big']
    );
  });
});
