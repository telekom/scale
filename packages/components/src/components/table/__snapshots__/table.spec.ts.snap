// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match snapshot 1`] = `
<scale-table class="table-0-0-4">
  <!---->
  <h5 hidden="" slot="header">
    Table Header
  </h5>
  <style>
    .table-0-0-4 {}
.table-0-0-4 [slot="table"] {
  overflow: auto;
  border-radius: 4px 4px 0 0;
  background: #F5F5F5;
}
.table-0-0-4 table {
  width: 100%;
  white-space: nowrap;
  border-spacing: 0;
  border-collapse: collapse;
  font-family: 'Roboto', 'Helvetica', 'Arial', sans-serif;
}
.table-0-0-4 th {
  font-size: 12px;
  text-align: left;
  line-height: 32px;
  user-select: none;
  color: #757575;
}
.table-0-0-4 th, .table-0-0-4 td {
  padding: 0 8px;
}
.table-0-0-4 tfoot, .table-0-0-4 tbody {
  background: #FFFFFF;
}
.table-0-0-4 tfoot tr td {
  padding: 8px;
  font-weight: 600;
  border-bottom: 1px solid #000000;
}
.table--variant-compressed-0-0-5 {}
.table--variant-compressed-0-0-5 td {
  padding: 8px;
}
.table--sortable-0-0-6 {}
.table--sortable-0-0-6 th {
  cursor: pointer;
}
.table--sortable-0-0-6 th[aria-disabled="true"] {
  pointer-events: none;
}
.table--sortable-0-0-6 th[aria-sort="descending"] .scale-sort-indicator polygon:first-of-type {
  fill: #000000;
}
.table--sortable-0-0-6 th[aria-sort="descending"] .scale-sort-indicator polygon {
  fill: #DCDCDC;
}
.table--sortable-0-0-6 th[aria-sort="ascending"] .scale-sort-indicator polygon:first-of-type {
  fill: #DCDCDC;
}
.table--sortable-0-0-6 th[aria-sort="ascending"] .scale-sort-indicator polygon {
  fill: #000000;
}
.table--sortable-0-0-6 th .scale-sort-indicator {
  left: -8px;
  width: 12px;
  height: 20px;
  display: inline-block;
  position: relative;
}
.table--sortable-0-0-6 th .scale-sort-indicator > svg {
  top: 4px;
  left: 0;
  width: 24px;
  height: 24px;
  position: absolute;
}
.table--sortable-0-0-6 th .scale-sort-indicator polygon {
  fill: #DCDCDC;
}
  </style>
  <table slot="table">
    <thead>
      <tr>
        <th aria-sort="descending">
          Title
        </th>
        <th>
          Time
        </th>
        <th>
          Euros
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          University of Plymouth
        </td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
      <tr>
        <td>
          University of Plymouth
        </td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
      <tr>
        <td>
          University of Plymouth
        </td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td>
          Total
        </td>
        <td></td>
        <td></td>
        <td>
          00:00:20
        </td>
        <td>
          100.245,10
        </td>
      </tr>
    </tfoot>
  </table>
</scale-table>
`;
